### PR TITLE
drivers: atmel_wdt: upgrade to support sama7g5 watchdog

### PR DESCRIPTION
**Pull request by TonyHan11**
In sama7g5 there's a DWDT (Dual Watchdog Timer) and the registers are not the same as the wdt for sama5d2. Here the DWD is handled as 2 watchdogs.

### CODE REVIEW

File: core/drivers/atmel_wdt.c [hunk 1]
```diff
@@ -283,8 +365,27 @@ static TEE_Result wdt_node_probe(const void *fdt, int node,
...
+static const struct wdt_compat sama5d2_compat = {
+	.wdt_ps = true,
+	.type = WDT_TYPE_WDT,
+	.dis_mask = WDT_MR_WDDIS,
+};
+
+static const struct wdt_compat sama7g5_compat = {
+	.wdt_ps = false,
+	.type = WDT_TYPE_DWDT,
+	.dis_mask = BIT(12),
```
Comment (jforissier) [inline core/drivers/atmel_wdt.c:377]: It would be nice to have a define for this bit

File: core/drivers/atmel_wdt.c [hunk 2]
```diff
@@ -11,12 +11,10 @@
 #include <kernel/delay.h>
 #include <kernel/dt.h>
 #include <kernel/dt_driver.h>
-#include <kernel/interrupt.h>
 #include <kernel/pm.h>
 #include <matrix.h>
 #include <mm/core_mmu.h>
 #include <platform_config.h>
-#include <tee_api_types.h>
```
Comment (jforissier) [inline core/drivers/atmel_wdt.c:19]: Removing these #include should be a separate commit

File: core/drivers/atmel_wdt.c [hunk 3]
```diff
@@ -59,12 +89,24 @@
...
+	enum wdt_type type; /* Type of Watchdog Timer */
+	uint32_t dis_mask; /* Mask of Watchdog Disable in Mode Register */
+};
 
 struct atmel_wdt {
 	struct wdt_chip chip;
+	enum wdt_type type;
+	uint32_t dis_mask;
 	vaddr_t base;
-	unsigned long rate;
```
Comment (jforissier) [inline core/drivers/atmel_wdt.c:67]: This too should be a separate commit

**Review by jforissier — COMMENTED** [review]
Hello @TonyHan11,
Please see a couple of comments below. With that addressed:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by TonyHan11** [comment]
Updated with adding a definition for the `disable_mask`, adding 2 separate commits for removing the unused header file and variable, review tags added. Thank you very much @jforissier

File: core/drivers/atmel_wdt.c (hunk 2 above)
Comment (jforissier) [reply core/drivers/atmel_wdt.c:19]: `#include <kernel/interrupt.h>` should be removed in this commit also

Comment (TonyHan11) [reply core/drivers/atmel_wdt.c:19]: Do you mean the `kernel/interrupt.h` in [core/include/drivers/wdt.h](https://github.com/OP-TEE/optee_os/blob/63bfec5e2/core/include/drivers/wdt.h#L10)? As interrupt is used in [this file](https://github.com/OP-TEE/optee_os/blob/63bfec5e264b66c00bfc1b6ba807ec0f75495891/core/drivers/atmel_wdt.c#L252) the header can not be removed.
Updated with removing some other header files, thanks.

File: core/drivers/atmel_wdt.c [hunk 4]
```diff
@@ -224,8 +295,9 @@ static void atmel_wdt_register_pm(struct atmel_wdt *wdt __unused)
 #endif
 
 static TEE_Result wdt_node_probe(const void *fdt, int node,
-				 const void *compat_data __unused)
+				 const void *compat_data)
 {
+	struct wdt_compat *compat = (struct wdt_compat *)compat_data;
```
Comment (etienne-lms) [inline core/drivers/atmel_wdt.c:300]: could use `const struct wdt_compat *compat` and remove the cast.

**Comment by TonyHan11** [comment]
Removed the  `!` characters and update the code for `*compat`, thanks.

File: core/drivers/atmel_wdt.c (hunk 2 above)
Comment (etienne-lms) [reply core/drivers/atmel_wdt.c:19]: `#include <tee_api_types.h>` is needed for `TEE_Result` definition.
`#include <kernel/delay.h>` is needed for `udelay()` declaration.
`#include <kernel/dt_driver.h>` is needed for `DEFINE_DT_DRIVER()` definition.

**Review by etienne-lms — COMMENTED** [review]
Comments for commit "drivers: atmel_wdt: remove the header files not used from "#include"".
By the way, I don't think you need to list the removed header files in the commit message, but that's not a strong opinion.

File: core/drivers/atmel_wdt.c (hunk 2 above)
Comment (etienne-lms) [reply core/drivers/atmel_wdt.c:19]: I forget to save many other comments when posting my review outcome. Here they are.

Many of these header files are needed IMO:
`#include <kernel/delay.h>` is needed for `udelay()` declaration.
`#include <kernel/dt.h>` is needed for `dt_map_dev()` and `fdt_get_status()` declaration and some other macros.
`#include <kernel/dt_driver.h>` is needed for `DEFINE_DT_DRIVER()` and few other definitions/declarations.
`#include <mm/core_mmu.h>` is needed for `core_mmu_remove_mapping()` declaration.
`#include <tee_api_types.h>` could be omitted but _tee_api_defines.h_ is needed for definition of `TEE_SUCCESS` and `TEE_ERROR_*`.

`#include <kernel/interrupt.h>` is needed for many function declarations and types definitions.

There are some header files that are missing: _util.h_, _stdbool.h_, _stdint.h_, _kernel/panic.h_, _trace.h_, _malloc.h_.

**Review by etienne-lms — COMMENTED** [review]
Comments for commit "drivers: atmel_wdt: remove the header files not used from "#include"".
By the way, I don't think you need to list the removed header files in the commit message, but that's not a strong opinion.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"drivers: atmel_wdt: upgrade to support sama7g5 watchdog" and
"drivers: atmel_wdt: remove the unused variable from "struct atmel_wdt"".

File: core/drivers/atmel_wdt.c (hunk 2 above)
Comment (TonyHan11) [reply core/drivers/atmel_wdt.c:19]: Update `#include list of the header files by:
 - Remove the unused header files from "#include".
 - `#include` the header files needed explicitly even if they are included indirectly.

**Comment by TonyHan11** [comment]
Not remove `#include <kernel/interrupt.h>` from commit `drivers: atmel_wdt: upgrade to support sama7g5 watchdog`.

**Comment by TonyHan11** [comment]
Reabse the commits onto the latest 'master' branch.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>`

**Comment by TonyHan11** [comment]
`Acked-by` added.
